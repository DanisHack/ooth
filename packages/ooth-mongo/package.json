{
  "name": "ooth-mongo",
  "version": "1.2.1",
  "description": "",
  "main": "lib/index.js",
  "scripts": {
    "link": "npm run prepublish && npm run link",
    "prepublish": "npm run build",
    "clean": "del-cli ./lib ./coverage",
    "build": "npm run clean && tsc -p ./config/tsconfig.build.json",
    "lint": "tslint -c ./tslint.json -p ./config/tsconfig.build.json",
    "test": "npm run lint && npm run clean && jest -c ./jest.json", 
    "test:watch": "npm run clean && jest -c ./jest.json --watch"
  },
  "author": "Nicola Marcacci Rossi <nicolamr@gmail.com> (http://nmr.io)",
  "license": "MIT",
  "devDependencies": {
    "@smartive/tslint-config": "^4.0.0",
    "@types/jest": "^23.3.1",
    "jest": "^21.2.1",
    "mongodb": "^2.2.33",
    "ts-jest": "^23.1.4",
    "tslint": "^5.11.0",
    "tsutils": "^3.0.0",
    "typescript": "^3.0.1"
  }
}
