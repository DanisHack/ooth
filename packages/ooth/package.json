{
  "name": "ooth",
  "version": "1.5.1",
  "description": "",
  "main": "lib/index.js",
  "scripts": {
    "link": "yarn prepublish && yarn link",
    "prepublish": "yarn build",
    "build": "babel src --out-dir lib --ignore '**/*.test.js'",
    "test": "jest"
  },
  "author": "Nick Redmark <nicolamr@gmail.com> (http://nmr.io)",
  "license": "MIT",
  "dependencies": {
    "body-parser": "^1.16.0",
    "cookie-parser": "^1.4.3",
    "express": "^4.14.1",
    "express-session": "^1.15.0",
    "express-ws": "^3.0.0",
    "jsonwebtoken": "^7.2.1",
    "locale": "^0.1.0",
    "morgan": "^1.7.0",
    "nodeify": "^1.0.0",
    "ooth-i18n": "^1.5.0",
    "passport": "^0.3.2",
    "passport-jwt": "^2.2.1"
  },
  "devDependencies": {
    "jest": "^20.0.4",
    "mongodb": "^2.2.31",
    "ooth-mongo": "^1.2.1",
    "passport-custom": "^1.0.5",
    "passport-strategy": "^1.0.0",
    "request-promise": "^4.2.1"
  }
}
