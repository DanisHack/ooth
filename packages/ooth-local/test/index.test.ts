import * as express from 'express';
import * as session from 'express-session';
import { cloneDeep } from 'lodash';
import { MongoClient } from 'mongodb';
import MongodbMemoryServer from 'mongodb-memory-server';
import { Ooth } from 'ooth';
import { OothMongo } from 'ooth-mongo';
import * as request from 'request-promise';

import oothLocal from '../src';

let mongoServer;
let con;
let app;
let server;
let ooth: Ooth;
let db;
let cookies = '';
let oothMongo: OothMongo;

const startServer = () => {
  return new Promise((resolve) => {
    server = app.listen(8080, resolve());
  });
};

const obfuscate = (obj, ...paths) => {
  const res = cloneDeep(obj);
  for (const path of paths) {
    const keys = path.split('.');
    let current = res;
    for (const key of keys.slice(0, -1)) {
      current = current[key];
    }
    current[keys[keys.length - 1]] = '<obfuscated>';
  }

  return res;
};

describe('ooth-local', () => {
  let onForgotPasswordListener;
  let onRequestVerifyListener;
  let onRegisterListener;
  let userId;
  let resetToken;
  let verificationToken;

  beforeAll(async () => {
    mongoServer = new MongodbMemoryServer();
    const connectionString = await mongoServer.getConnectionString();
    const dbName = await mongoServer.getDbName();
    con = await MongoClient.connect(connectionString);
    db = con.db(dbName);
  });

  afterAll(async () => {
    await con.close();
    await mongoServer.stop();
  });

  beforeEach(async () => {
    app = express();
    app.use(
      session({
        name: 'api-session-id',
        secret: 'x',
        resave: false,
        saveUninitialized: true,
      }),
    );
    oothMongo = new OothMongo(db);
    ooth = new Ooth({
      app,
      backend: oothMongo,
      sharedSecret: '',
      standalone: false,
      path: '',
      onLogin: () => null,
      onLogout: () => null,
    });
    oothLocal({
      ooth,
      onRegister(...args) {
        if (onRegisterListener) {
          onRegisterListener(...args);
        }
      },
      onForgotPassword(data) {
        if (onForgotPasswordListener) {
          //Store user id and reset token for follow on test
          userId = data._id;
          resetToken = data.passwordResetToken;
          onForgotPasswordListener(data);
        }
      },
      onGenerateVerificationToken(data) {
        if (onRequestVerifyListener) {
          //Store user id and  verification token for follow on test
          userId = data._id;
          verificationToken = data.verificationToken;
          onRequestVerifyListener(data);
        }
      },
      translations: {
        en: require('../i18n/en.json'),
        fr: {
          register: {
            invalid_email: 'french invalid email',
          },
        },
      },
    });
    await startServer();
  });

  afterEach(async () => {
    if (server) {
      await server.close();
    }
    if (db) {
      await db.dropDatabase();
    }
    cookies = '';
  });

  test('can check status', async () => {
    const res = await request({
      uri: 'http://localhost:8080/status',
      json: true,
    });
    expect(res).toMatchSnapshot();
  });

  test('fails without email', async () => {
    try {
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('translates', async () => {
    try {
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        headers: {
          'Accept-Language': 'fr',
        },
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('fails without password', async () => {
    try {
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        body: {
          email: 'test@example.com',
        },
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('fails without valid password', async () => {
    try {
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        body: {
          email: 'test@example.com',
          password: 'xxxxxx',
        },
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('can register', async () => {
    onRegisterListener = jest.fn();
    const res = await request({
      method: 'POST',
      uri: 'http://localhost:8080/local/register',
      body: {
        email: 'test@example.com',
        password: 'Asdflba09',
      },
      json: true,
    });
    expect(res).toMatchSnapshot();
    expect(obfuscate(onRegisterListener.mock.calls[0][0], '_id', 'verificationToken')).toMatchSnapshot();
  });

  describe('after registration', () => {
    beforeEach(async () => {
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        body: {
          email: 'test@example.com',
          password: 'Asdflba09',
        },
        json: true,
      });
    });

    test('login fails with wrong password', async () => {
      try {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/login',
          body: {
            username: 'test@example.com',
            password: 'Asdflba099',
          },
          json: true,
        });
      } catch (e) {
        expect(e.response.body).toMatchSnapshot();
        return;
      }
      throw new Error("Didn't fail");
    });

    test('can login', async () => {
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/login',
        body: {
          username: 'test@example.com',
          password: 'Asdflba09',
        },
        json: true,
      });
      delete res.user._id;
      expect(res).toMatchSnapshot();
    });

    test('can forget password', async () => {
      onForgotPasswordListener = jest.fn();
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/forgot-password',
        body: {
          username: 'test@example.com',
        },
        json: true,
      });
      expect(obfuscate(onForgotPasswordListener.mock.calls[0][0], '_id', 'passwordResetToken')).toMatchSnapshot();
    });

    test('can reset password', async () => {
      onForgotPasswordListener = jest.fn();

      //Store the password reset token and userId so we can use it later
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/forgot-password',
        body: {
          username: 'test@example.com',
        },
        json: true,
      });
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/reset-password',
        body: {
          userId,
          token: resetToken,
          newPassword: 'Asdflba10',
        },
        json: true,
      });
      expect(res).toMatchSnapshot();
    });

    test("can't reset password with invalid token", async () => {
      onForgotPasswordListener = jest.fn();

      //Store the password reset token and userId so we can use it later
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/forgot-password',
        body: {
          username: 'test@example.com',
        },
        json: true,
      });
      try {
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/reset-password',
          body: {
            userId,
            token: 'wrongtoken',
            newPassword: 'Asdflba10',
          },
          json: true,
        });
      } catch (e) {
        expect(e.response.body).toMatchSnapshot();
        return;
      }
      throw new Error("Didn't fail");
    });

    describe('after login', () => {
      beforeEach(async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/login',
          body: {
            username: 'test@example.com',
            password: 'Asdflba09',
          },
          json: true,
          resolveWithFullResponse: true,
        });
        cookies = res.headers['set-cookie'];
      });

      test('can generate verification token', async () => {
        onRequestVerifyListener = jest.fn();
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/generate-verification-token',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        expect(obfuscate(onRequestVerifyListener.mock.calls[0][0], '_id', 'verificationToken')).toMatchSnapshot();
      });

      test('can verify user with token', async () => {
        onRequestVerifyListener = jest.fn();
        //Store verification token and userId for later use
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/generate-verification-token',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/verify',
          json: true,
          body: {
            userId,
            token: verificationToken,
          },
          headers: {
            Cookie: cookies,
          },
        });
        expect(obfuscate(res, 'user._id')).toMatchSnapshot();
      });

      test("can't verify user with invalid token", async () => {
        onRequestVerifyListener = jest.fn();
        //Store verification token and userId for later use
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/generate-verification-token',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/local/verify',
            json: true,
            body: {
              userId,
              token: 'wrongtoken',
            },
            headers: {
              Cookie: cookies,
            },
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't fail");
      });

      test('can check status', async () => {
        const res = await request({
          uri: 'http://localhost:8080/status',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        delete res.user._id;
        expect(res).toMatchSnapshot();
      });

      test('can set username', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/set-username',
          body: {
            username: 'heythere12_',
          },
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        delete res.user._id;
        expect(res).toMatchSnapshot();
      });

      test('can set email', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/set-email',
          body: {
            email: 'test@example.com',
          },
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        delete res.user._id;
        expect(res).toMatchSnapshot();
      });

      test("can't set invalid username", async () => {
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/local/set-username',
            body: {
              username: 'bl',
            },
            json: true,
            headers: {
              Cookie: cookies,
            },
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't fail");
      });
    });
  });
});
