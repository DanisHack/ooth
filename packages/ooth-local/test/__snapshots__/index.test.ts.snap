// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ooth-local after registration after login can check status 1`] = `
Object {
  "user": Object {
    "local": Object {
      "email": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can generate verification token 1`] = `
Object {
  "_id": "<obfuscated>",
  "email": "test@example.com",
  "verificationToken": "<obfuscated>",
}
`;

exports[`ooth-local after registration after login can set email 1`] = `
Object {
  "message": "Email updated.",
  "user": Object {
    "local": Object {
      "email": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can set username 1`] = `
Object {
  "message": "Username updated.",
  "user": Object {
    "local": Object {
      "email": "test@example.com",
      "username": "heythere12_",
    },
  },
}
`;

exports[`ooth-local after registration after login can verify user with token 1`] = `
Object {
  "message": "Email verified",
  "user": Object {
    "_id": "<obfuscated>",
    "local": Object {
      "email": "test@example.com",
      "verified": true,
    },
  },
}
`;

exports[`ooth-local after registration after login can't set invalid username 1`] = `
Object {
  "message": "Username must be all lowercase, contain only letters, numbers and _ (starting with a letter), and be between 4 and 20 characters long.",
  "status": "error",
}
`;

exports[`ooth-local after registration after login can't verify user with invalid token 1`] = `
Object {
  "message": "Verification token invalid, expired or already used.",
  "status": "error",
}
`;

exports[`ooth-local after registration can forget password 1`] = `
Object {
  "_id": "<obfuscated>",
  "email": "test@example.com",
  "passwordResetToken": "<obfuscated>",
}
`;

exports[`ooth-local after registration can login 1`] = `
Object {
  "user": Object {
    "local": Object {
      "email": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration can reset password 1`] = `
Object {
  "message": "Password has been reset.",
}
`;

exports[`ooth-local after registration can't reset password with invalid token 1`] = `
Object {
  "message": "Password reset token invalid, expired or already used.",
  "status": "error",
}
`;

exports[`ooth-local after registration login fails with wrong password 1`] = `
Object {
  "message": "Incorrect password.",
  "status": "error",
}
`;

exports[`ooth-local can check status 1`] = `
Object {
  "user": null,
}
`;

exports[`ooth-local can register 1`] = `
Object {
  "message": "User registered successfully.",
}
`;

exports[`ooth-local can register 2`] = `
Object {
  "_id": "<obfuscated>",
  "email": "test@example.com",
  "verificationToken": "<obfuscated>",
}
`;

exports[`ooth-local fails without email 1`] = `
Object {
  "message": "Invalid email",
  "status": "error",
}
`;

exports[`ooth-local fails without password 1`] = `
Object {
  "message": "Invalid password",
  "status": "error",
}
`;

exports[`ooth-local fails without valid password 1`] = `
Object {
  "message": "Password must contain digits, lowercase and uppercase letters and be at least 6 characters long.",
  "status": "error",
}
`;

exports[`ooth-local translates 1`] = `
Object {
  "message": "Invalid email",
  "status": "error",
}
`;
